safe/gis/tools.py

```diff
@@ -113,45 +129,194 @@ def reclassify_value(one_value, ranges):
+    :rtype: tuple
+    """
+    if not full_layer_uri_string:
+        return None, None
+
+    split = full_layer_uri_string.split('|qgis_provider=')
+    if len(split) == 1:
+        return split[0], None
+    else:
+        return split
+
+
+def load_layer(full_layer_uri_string, name=None, provider=None):
+    """Helper to load and return a single QGIS layer based on our layer URI.
+
+    :param provider:
+    :type provider:
+
+    :param name: The name of the layer. If not provided, it will be computed
+        based on the URI.
+    :type name: basestring
+
+    :param full_layer_uri_string: Layer URI, with provider type.
+    :type full_layer_uri_string: str
 
     :returns: tuple containing layer and its layer_purpose.
     :rtype: (QgsMapLayer, str)
     """
-    # Extract basename and absolute path
-    file_name = os.path.split(layer_path)[-1]  # In case path was absolute
-    base_name, extension = os.path.splitext(file_name)
+    if provider:
+        # Cool !
+        layer_path = full_layer_uri_string
+    else:
+        #  Let's check if the driver is included in the path
+        layer_path, provider = decode_full_layer_uri(full_layer_uri_string)
+
+        if not provider:
+            # Let's try to check if it's file based and look for a extension
+            if '|' in layer_path:
+                clean_uri = layer_path.split('|')[0]
+            else:
+                clean_uri = layer_path
+            is_file_based = os.path.exists(clean_uri)
+            if is_file_based:
+                # Extract basename and absolute path
+                file_name = os.path.split(layer_path)[
+                    -1]  # If path was absolute
+                extension = os.path.splitext(file_name)[1]
+                if extension in OGR_EXTENSIONS:
+                    provider = 'ogr'
+                elif extension in GDAL_EXTENSIONS:
+                    provider = 'gdal'
+                else:
+                    provider = None
+
+    if not provider:
+        layer = load_layer_without_provider(layer_path)
+    else:
+        layer = load_layer_with_provider(layer_path, provider)
+
+    if not layer or not layer.isValid():
+        message = 'Layer "%s" is not valid' % layer_path
+        LOGGER.debug(message)
+        raise Exception(message)
+
+    # Define the name
+    if not name:
+        source = layer.source()
+        if '|' in source:
+            clean_uri = source.split('|')[0]
+        else:
+            clean_uri = source
+        is_file_based = os.path.exists(clean_uri)
+        if is_file_based:
+            # Extract basename and absolute path
+            file_name = os.path.split(layer_path)[-1]  # If path was absolute
+            name = os.path.splitext(file_name)[0]
+        else:
+            # Might be a DB, take the DB name
+            source = QgsDataSourceURI(source)
+            name = source.table()
+
+    if not name:
+        name = 'default'
+
+    if qgis_version() >= 21800:
```

Comment:
IIRC, you told me that we don't need to do this? Or is it the other API?

Reply:
The getter is the same `name()` whatever the QGIS version, but the setter is different between QGIS 2.14 and 2.18. So it's fine.